### PR TITLE
Controllers

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR introduces a set of classes, which encapsulate the logic of data loading, model training, inference and evaluation of results:

- Detector: a base class that allows vanilla horizon detection

- Enhancer: improve quality of labeled points

- Extender: increase area covered by a predicted horizon

- Extractor: get multiple horizons for a cube in a completely unsupervised way

### CODE REVIEW

File: seismiqb/src/controllers/enhancer.py [hunk 1]
```diff
@@ -107,10 +107,8 @@ def inference(self, horizon, **kwargs):
         dataset = self._make_dataset(horizon)
         super().inference(dataset, **kwargs)
 
-    def get_train_template(self):
-        """ Define training pipeline.
-
-        Following parameters are fetched from pipeline config: `model_config`.
+    def get_load_ppl(self):
```
Comment (roman-kh) [inline seismiqb/src/controllers/enhancer.py:110]: No need for abbreviation - write `pipeline` in full.
`get` can be removed though

File: seismiqb/src/controllers/best_practices.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import torch
+import torch.nn as nn
+
+from ...batchflow.batchflow.models.torch import ResBlock
+
+class Dice(nn.Module):
+    """ Dice coefficient as a loss function. """
+    def forward(self, prediction, target):
+        prediction = torch.sigmoid(input)
```
Comment (dmylzenova) [inline seismiqb/src/controllers/best_practices.py:11]: prediction = torch.sigmoid(prediction)

File: seismiqb/src/controllers/enhancer.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            .load_cubes(dst='images')
+            .adaptive_reshape(src=['images', 'masks'],
+                              shape=self.crop_shape)
+            .scale(mode='q', src='images')
+        )
+
+    def distortion_pipeline(self):
+        """ Defines transformations performed with `masks` component. """
+        def binarize(batch):
+            batch.prior_masks = (batch.prior_masks > 0).astype(int).astype(np.float32)
```
Comment (dmylzenova) [inline seismiqb/src/controllers/enhancer.py:75]: `batch.prior_masks = (batch.prior_masks > 0).astype(np.float32)` is enough

File: seismiqb/src/controllers/enhancer.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        """ Run all steps of the Enhancement procedure including creating dataset for
+        the given horizon, creating instance of the class and running train and inference
+        methods.
+
+        Parameters
+        ----------
+        horizon : an instance of :class:`.Horizon`
+            A horizon to be extended
+        n_steps : int
+            Number of steps of the Extension algorithm.
```
Comment (dmylzenova) [inline seismiqb/src/controllers/enhancer.py:164]: not a parameter here

File: seismiqb/src/controllers/extender.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
+""" A holder for horizon extension steps inherited from `.class:Enhancer` with:
```
Comment (dmylzenova) [inline seismiqb/src/controllers/extender.py:1]: Add more description

File: seismiqb/src/controllers/extender.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    - redifined get_mask_transform_ppl to filter out a true mask to train an extension model
+      on a horizon area with a given percentage of holes.
+    - making inference of a Horizon Extension algorithm to cover the holes in a given horizon.
```
Comment (dmylzenova) [inline seismiqb/src/controllers/extender.py:4]: what algorithm? clarify

File: seismiqb/src/controllers/extender.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        -------
+        Extended horizon.
+        """
+        model_config = MODEL_CONFIG_EXTENSION if model_config is None else model_config
+        extender = Extender(save_dir=save_dir, model_config=model_config, device=device,
+                            crop_shape=crop_shape, batch_size=batch_size)
+        extender.train(horizon, n_iters=n_iters, use_grid=False)
+        extended = extender.inference(horizon, n_steps=n_steps,
+                                      batch_size=batch_size, stride=stride)
+        return extended
```
Comment (dmylzenova) [inline seismiqb/src/controllers/extender.py:206]: behavior differs from Enhancer.run

File: seismiqb/src/controllers/base.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+        msg = '\n'
+        for idx in dataset.indices:
+            msg += f'{idx}\n'
+            for hor in dataset.labels[idx]:
+                msg += f'    {hor.name}'
+        self.log(f'Created dataset ::: {msg}')
+        return dataset
+
+
+    def make_dataset_from_horizon(self, horizon):
```
Comment (dmylzenova) [inline seismiqb/src/controllers/base.py:148]: these dataset creation methods can be combined

File: seismiqb/src/controllers/base.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+            crop_shape_grid = np.array(self.crop_shape)[[1, 0, 2]]
+            config['side_view'] = 1.0
+            config['order'] = (1, 0, 2)
+        strides_grid = [max(1, int(item//overlap_factor))
+                        for item in crop_shape_grid]
+        return config, crop_shape_grid, strides_grid
+
+
+    def inference_0(self, dataset, heights_range=None, orientation='i', overlap_factor=2, **kwargs):
+        """ Inference on chunks, assemble into massive 3D array, extract horizon surface. """
```
Comment (dmylzenova) [inline seismiqb/src/controllers/base.py:407]: documentation on inference_0 and inference_1 can be more descriptive. How can one choose between them?

File: seismiqb/src/horizon.py [hunk 10]
```diff
@@ -1703,6 +1703,16 @@ def show_slide(self, loc, width=3, axis='i', order_axes=None, zoom_slice=None, *
...
 
 
+    def copy(self):
+        """ Create a shallow copy of a horizon.
+
+        Returns
+        -------
+        A horizon object with new matrix object and a reference to the old geometry attribute.
+        """
+        return Horizon(np.copy(self.matrix), self.geometry, i_min=self.i_min, x_min=self.x_min)
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:1713]: Maybe add `name` to it?

File: seismiqb/src/controllers/extender.py [hunk 11]
```diff
@@ -76,18 +76,19 @@ def inference(self, horizon, n_steps=30, batch_size=128, stride=16):
             'batch_size': batch_size,
             'model_pipeline': self.model_pipeline
         }
-        prev_len = len(horizon)
+        pred_horizon = horizon.copy()
```
Comment (SergeyTsimfer) [inline seismiqb/src/controllers/extender.py:79]: If you don't need the `horizon` variable anymore, maybe simple `horizon = horizon.copy()` would suffice?

File: seismiqb/src/horizon.py [hunk 12]
```diff
@@ -1703,6 +1703,16 @@ def show_slide(self, loc, width=3, axis='i', order_axes=None, zoom_slice=None, *
         plot_image([seismic_slide, mask], order_axes=order_axes, **kwargs)
 
 
+    def copy(self):
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:1706]: Maybe, `__copy__`?

File: models/Horizon extrapolation/Horizon_extension.ipynb [hunk 13]
```diff
@@ -0,0 +1,765 @@
+{
```
Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>As you can see where..?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Elaborate, if possible</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Elaborate on this: give links to previous notebooks, to tutorials, etc</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>If the line is not needed, then there is no need in it</p><p>Let's give our datasets some respect by naming their variables, well, <code>dataset</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>That is a nice name of a horizon</p><p> </p><p>Add the metric picture</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Explain what these are'</p><p>Make <code>CROP_SHAPE</code>  tuple</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>That is one of a flighty sentence</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>The data is already loaded; that is not why you need this part</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>If the line is not needed... you know the wibe</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Why is this standard? By whom was it standardized?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>...modify by an additional encoding branch with prior information about the horizon of interest. Information this new branch is repeatedly incorporated into the main branch before each of the downsampling steps.</p><p>... but you can find our PyTorch implementation in &lt;&gt;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Looking cool, </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Tell about the output stide; optimizer, LR, decay</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>should not it be defined with other global constants?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Extension</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>A very sketchy description of what is going on is intervened with quite technical details, which is confusing</p><p>Pretty much every of your <code>pipeline description cells</code>  would benefit from bulleted lists usage</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Looks like a global constant to me</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>Make this bar behave</p><p>STRIDE looks like a global constant to me</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (SergeyTsimfer) [inline models/Horizon extrapolation/Horizon_extension.ipynb:1]: <p>to evaluate we evaluate</p><p>makes sense!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gazprom-neft/seismiqb/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (dmylzenova) [reply models/Horizon extrapolation/Horizon_extension.ipynb]: <p>output stide?</p>
